File: LogAndroid/TrustarApp/src/main/kotlin/jp/co/toukei/log/trustar/rest/response/RawChart.kt
package jp.co.toukei.log.trustar.rest.response

import annotation.JsonObjectExtra
import annotation.Keep
import annotation.Para
import jp.co.toukei.log.trustar.common.StoredFile
import jp.co.toukei.log.trustar.common.asDbStoredFile
import jp.co.toukei.log.trustar.db.allocationRowOrNull
import jp.co.toukei.log.trustar.db.user.entity.DeliveryChart
import org.json.JSONObject

class RawChart @Keep constructor(
    @Para("chartHeaders") @JvmField val chartHeaders: Array<RawChartHeader>?,
    @Para("chartMemos") @JvmField val chartMemos: Array<RawChartMemo>?,
    @Para("chartFiles") @JvmField val chartFiles: Array<RawChartFile>?,
) {
    @JvmField
    val charts = run {
        val memeMap = chartMemos?.groupBy { it.chartCd }
        val fMap = chartFiles?.groupBy { it.chartCd }
        chartHeaders?.map { chartHeader ->
            DeliveryChart(
                chartCd = chartHeader.chartCd,
                placeCd = chartHeader.placeCd,
                info = DeliveryChart.Info(
                    dest = chartHeader.chartNm.orEmpty(),
                    addr1 = chartHeader.addr1.orEmpty(),
                    addr2 = chartHeader.addr2.orEmpty(),
                    tel = chartHeader.tel.orEmpty(),
                    carrier = chartHeader.carrier.orEmpty(),
                    carrierTel = chartHeader.carrierTel.orEmpty(),
                ),
                memos = memeMap?.get(chartHeader.chartCd)?.map { m ->
                    DeliveryChart.ChartMemo(
                        label = m.title.orEmpty(),
                        note = m.note.orEmpty(),
                        highlight = m.highLightFlag,
                        extra = m.extraJson
                    )
                } ?: emptyList(),
                images = fMap?.get(chartHeader.chartCd)?.map { f ->
                    DeliveryChart.ChartImageFile(
                        dbStoredFile = StoredFile.ByKey(f.fileKey).asDbStoredFile(),
                        extra = f.extraJson,
                    )
                } ?: emptyList(),
                lastAllocationRow = allocationRowOrNull(
                    chartHeader.allocationNo,
                    chartHeader.allocationRowNo
                ),
                extra = chartHeader.extraJson,
            )
        } ?: emptyList()
    }
}


class RawChartHeader @Keep constructor(
    @Para("chartCd") @JvmField val chartCd: String,
    @Para("placeCd") @JvmField val placeCd: String?,
    @Para("chartNm") @JvmField val chartNm: String?,
    @Para("chartAddr1") @JvmField val addr1: String?,
    @Para("chartAddr2") @JvmField val addr2: String?,
    @Para("chartTel") @JvmField val tel: String?,
    @Para("chargeNm") @JvmField val carrier: String?,
    @Para("chargeTel") @JvmField val carrierTel: String?,
    @Para("allocationNo") @JvmField val allocationNo: String?,
    @Para("allocationRowNo") @JvmField val allocationRowNo: Int?,

    // wtf???
    @JsonObjectExtra("mkTimeStamp", "mkUserId", "mkPrgNm") @JvmField val extraJson: JSONObject,
)

class RawChartMemo @Keep constructor(
    @Para("ChartCd") @JvmField val chartCd: String,
    @Para("HighLightFlag") highLightFlag: Int,
    @Para("Title") @JvmField val title: String?,
    @Para("Note") @JvmField val note: String?,

    // wtf???
    @JsonObjectExtra("MkTimeStamp", "MkUserId", "MkPrgNm") @JvmField val extraJson: JSONObject,
) {
    @JvmField
    val highLightFlag = highLightFlag == 1
}

class RawChartFile @Keep constructor(
    @Para("ChartCd") @JvmField val chartCd: String,
    @Para("FileId") @JvmField val fileKey: String,

    // wtf???
    @JsonObjectExtra("MkTimeStamp", "MkUserId", "MkPrgNm") @JvmField val extraJson: JSONObject,
)

File: LogiOS/TrustarApp/Services/Response/RawChart.swift
//
//  DeliveryChartResult.swift
//  TrustarApp
//
//  Created by hoanx on 8/9/23.
//  Copyright © 2023 DionSoftware. All rights reserved.
//
import Foundation

// MARK: - Welcome
class RawChart: Codable {
    let binHeaders, binResults, binDetails, binCollections: String?
    let binCharts: BinCharts
    let workResults: String?
    let clientInfo: String?

    init(binHeaders: String?, binResults: String?, binDetails: String?, binCollections: String?, binCharts: BinCharts,
         workResults: String?, clientInfo: String?) {
        self.binHeaders = binHeaders
        self.binResults = binResults
        self.binDetails = binDetails
        self.binCollections = binCollections
        self.binCharts = binCharts
        self.workResults = workResults
        self.clientInfo = clientInfo
    }
    
    func charts() -> [DeliveryChart]? {
        var chartArr:[DeliveryChart] = [];
        guard let chartHeader = self.binCharts.chartHeaders else {
            return chartArr
        }

       for item in chartHeader {
           do {
               var memos: [ChartMemos] = []
               var imgs: [ChartImageFile] = []
               if let chartMemos = self.binCharts.chartMemos?.filter({$0.chartCD == item.chartCD}) {
                   for mm in chartMemos {
                       let extraMM = "{\"MkTimeStamp\":\"\(mm.mkTimeStamp ?? "")\",\"MkUserId\":\"\(mm.mkUserID ?? "")\",\"MkPrgNm\":\"\(mm.mkPrgNm ?? "")\"}"
                       memos.append(ChartMemos(label: mm.title ?? "", note: mm.note ?? "",
                                               higtlight: mm.highLightFlag == 1, extra: extraMM))
                   }
               }

               if let files = self.binCharts.chartFiles?.filter({$0.chartCD == item.chartCD}) {
                   for file in files {
                       let extraImg = "{\"MkTimeStamp\":\"\(file.mkTimeStamp ?? "")\",\"MkUserId\":\"\(file.mkUserID ?? "")\",\"MkPrgNm\":\"\(file.mkPrgNm ?? "")\"}"
                       imgs.append(ChartImageFile(
                        dbStoreFile: StoredFile.ByKey(key: file.fileID ?? "").asDbStoredFile(), extra: extraImg))
                   }
               }

               let jsonEn = JSONEncoder()
               let chart = DeliveryChart()
               chart.chartCd = item.chartCD
               chart.placeCd = item.placeCD ?? ""
               chart.dest = item.chartNm ?? ""
               chart.addr1 = item.chartAddr1 ?? ""
               chart.addr2 = item.chartAddr2 ?? ""
               chart.tel = item.chartTel ?? ""
               chart.carrier = item.chargeNm ?? ""
               chart.carrierTel = item.chargeTel ?? ""
               chart.memos = try String(data: jsonEn.encode(memos), encoding: .utf8)!
               chart.images = try String(data: jsonEn.encode(imgs), encoding: .utf8)!
               let extra = "{\"mkTimeStamp\":\"\(item.mkTimeStamp ?? "")\",\"mkUserId\":\"\(item.mkUserID ?? "")\",\"mkPrgNm\":\"\(item.mkPrgNm ?? "")\"}"
               chart.extra = extra
               chartArr.append(chart)
           } catch {
               print(error)
           }
       }
       return chartArr
   }
}

// MARK: - BinCharts
class BinCharts: Codable {
   let chartHeaders: [ChartHeader]?
   let chartMemos: [ChartMemo]?
   let chartFiles: [ChartFile]?
}

// MARK: - ChartHeader
class ChartHeader: Codable {
    let chartCD: String
    let mkTimeStamp, mkUserID, mkPrgNm, companyCD: String?
    let allocationNo, allocationRowNo, chartNm: String?
    let placeCD, chartZip, chartAddr1, chartAddr2: String?
    let chartTel, chargeNm, chargeTel: String?

    enum CodingKeys: String, CodingKey {
        case mkTimeStamp
        case mkUserID = "mkUserId"
        case mkPrgNm
        case companyCD = "companyCd"
        case allocationNo, allocationRowNo
        case chartCD = "chartCd"
        case chartNm
        case placeCD = "placeCd"
        case chartZip, chartAddr1, chartAddr2, chartTel, chargeNm, chargeTel
    }

    init(mkTimeStamp: String?, mkUserID: String?, mkPrgNm: String?, companyCD: String?, allocationNo: String?, allocationRowNo: String?, chartCD: String, chartNm: String?, placeCD: String?, chartZip: String?, chartAddr1: String?, chartAddr2: String?, chartTel: String?, chargeNm: String?, chargeTel: String?) {
        self.mkTimeStamp = mkTimeStamp
        self.mkUserID = mkUserID
        self.mkPrgNm = mkPrgNm
        self.companyCD = companyCD
        self.allocationNo = allocationNo
        self.allocationRowNo = allocationRowNo
        self.chartCD = chartCD
        self.chartNm = chartNm
        self.placeCD = placeCD
        self.chartZip = chartZip
        self.chartAddr1 = chartAddr1
        self.chartAddr2 = chartAddr2
        self.chartTel = chartTel
        self.chargeNm = chargeNm
        self.chargeTel = chargeTel
    }
}

// MARK: - ChartMemo
class ChartMemo: Codable {
    let chartCD: String
    let mkTimeStamp, mkUserID, mkPrgNm: String?
    let seqNo, highLightFlag: Int?
    let title, note: String?

    enum CodingKeys: String, CodingKey {
        case mkTimeStamp = "MkTimeStamp"
        case mkUserID = "MkUserId"
        case mkPrgNm = "MkPrgNm"
        case chartCD = "ChartCd"
        case seqNo = "SeqNo"
        case highLightFlag = "HighLightFlag"
        case title = "Title"
        case note = "Note"
    }

    init(mkTimeStamp: String?, mkUserID: String?, mkPrgNm: String?, chartCD: String, seqNo: Int?, highLightFlag: Int?, title: String?, note: String?) {
        self.mkTimeStamp = mkTimeStamp
        self.mkUserID = mkUserID
        self.mkPrgNm = mkPrgNm
        self.chartCD = chartCD
        self.seqNo = seqNo
        self.highLightFlag = highLightFlag
        self.title = title
        self.note = note
    }
}

// MARK: - ChartFile
class ChartFile: Codable {
    let chartCD: String
    let mkTimeStamp, mkUserID, mkPrgNm: String?
    let fileID: String?
    let dataClass: Int?
    let attachedFile: String?
    let fileName: String?
    let fileSize, width, height: Int?
    let folderPath, fileExtension: String?

    enum CodingKeys: String, CodingKey {
        case mkTimeStamp = "MkTimeStamp"
        case mkUserID = "MkUserId"
        case mkPrgNm = "MkPrgNm"
        case chartCD = "ChartCd"
        case fileID = "FileId"
        case dataClass = "DataClass"
        case attachedFile = "AttachedFile"
        case fileName = "FileName"
        case fileSize = "FileSize"
        case width = "Width"
        case height = "Height"
        case folderPath = "FolderPath"
        case fileExtension = "FileExtension"
    }
}




